pkg/tempo/remotewriteexporter/exporter.go

```diff
@@ -0,0 +1,231 @@
+package remotewriteexporter
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/agent/pkg/prom/instance"
+	"github.com/grafana/agent/pkg/tempo/contextkeys"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/storage"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/atomic"
+)
+
+const (
+	nameLabelKey  = "__name__"
+	sumSuffix     = "sum"
+	countSuffix   = "count"
+	bucketSuffix  = "bucket"
+	leStr         = "le"
+	infBucket     = "+Inf"
+	counterSuffix = "total"
+	noSuffix      = ""
+)
+
+type dataPoint interface {
+	LabelsMap() pdata.StringMap
+	Timestamp() pdata.TimestampUnixNano
+}
+
+type remoteWriteExporter struct {
+	done atomic.Bool
+	prom *instance.Instance
+
+	constLabels labels.Labels
+	namespace   string
+}
+
+func newRemoteWriteExporter(cfg *Config) (component.MetricsExporter, error) {
+	return &remoteWriteExporter{
+		done:        atomic.Bool{},
+		constLabels: cfg.ConstLabels,
+		namespace:   cfg.Namespace,
+	}, nil
+}
+
+func (e *remoteWriteExporter) Start(ctx context.Context, _ component.Host) error {
+	prom := ctx.Value(contextkeys.Prometheus).(*instance.Instance)
+	if prom == nil {
+		return fmt.Errorf("key does not contain a Prometheus instance")
+	}
+	e.prom = prom
+	return nil
+}
+
+func (e *remoteWriteExporter) Shutdown(_ context.Context) error {
+	e.done.Store(true)
+	return nil
+}
+
+func (e *remoteWriteExporter) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
+	if e.done.Load() {
+		return nil
+	}
+
+	app := e.prom.Appender(ctx)
+
+	rm := md.ResourceMetrics()
+	for i := 0; i < rm.Len(); i++ {
+		ilm := rm.At(i).InstrumentationLibraryMetrics()
+		for j := 0; j < ilm.Len(); j++ {
+			ms := ilm.At(j).Metrics()
+			for k := 0; k < ms.Len(); k++ {
+				switch m := ms.At(k); m.DataType() {
+				case pdata.MetricDataTypeDoubleSum, pdata.MetricDataTypeIntSum, pdata.MetricDataTypeDoubleGauge, pdata.MetricDataTypeIntGauge:
+					if err := e.processScalarMetric(app, m); err != nil {
+						return fmt.Errorf("failed to process metric %s", err)
+					}
+				case pdata.MetricDataTypeDoubleHistogram, pdata.MetricDataTypeIntHistogram:
+					if err := e.processHistogramMetrics(app, m); err != nil {
+						return fmt.Errorf("failed to process metric %s", err)
+					}
+				case pdata.MetricDataTypeDoubleSummary:
+					return fmt.Errorf("%s processing unimplemented", m.DataType())
+				default:
+					return fmt.Errorf("unsupported m data type %s", m.DataType())
+				}
+			}
+		}
+	}
+
+	return app.Commit()
+}
+
+func (e *remoteWriteExporter) processHistogramMetrics(app storage.Appender, m pdata.Metric) error {
```

Comment:
this is obviously quite complex. i say add some testing and ship it. we can confirm internally that these metrics are in sync with others.

Reply:
Added a few unit tests, although I'm not extremely happy with them. I'd like to work something like https://github.com/prometheus/compliance/tree/main/remote_write out to the exporter. Maybe in a future pass?